File: ScaleformUI_Csharp/Scaleforms/Minimap/Minimap.cs
using CitizenFX.Core;
using static CitizenFX.Core.Native.API;

namespace ScaleformUI.Scaleforms
{
    public static class MinimapOverlays
    {
        internal static Dictionary<int, KeyValuePair<string, string>> minimaps = new();
        internal static int overlay = 0;

        private static async Task Load()
        {
            overlay = AddMinimapOverlay("MINIMAP_LOADER.gfx");
            while (!HasMinimapOverlayLoaded(overlay)) await BaseScript.Delay(0);
            SetMinimapOverlayDisplay(overlay, 0f, 0f, 100f, 100f, 100f);
        }

        /// <summary>
        /// Adds a custom minimap in any place of the map specified
        /// </summary>
        /// <param name="textureDict">The texture dictionary</param>
        /// <param name="textureName">The texture name</param>
        /// <param name="x">X position</param>
        /// <param name="y">Y Position</param>
        /// <returns>The overlay ID in Dictionary</returns>
        public static async Task<int> AddOverlayToMinimap(string textureDict, string textureName, float x, float y, float width = -1, float height = -1, int alpha = -1, bool centered = false)
        {
            if (overlay == 0) await Load();

            if (!HasStreamedTextureDictLoaded(textureDict))
            {
                RequestStreamedTextureDict(textureDict, false);
                while (!HasStreamedTextureDictLoaded(textureDict)) await BaseScript.Delay(0);
            }

            CallMinimapScaleformFunction(overlay, "ADD_OVERLAY");
            ScaleformMovieMethodAddParamTextureNameString(textureDict);
            ScaleformMovieMethodAddParamTextureNameString(textureName);
            ScaleformMovieMethodAddParamFloat(x);
            ScaleformMovieMethodAddParamFloat(y);
            ScaleformMovieMethodAddParamFloat(width);
            ScaleformMovieMethodAddParamFloat(height);
            ScaleformMovieMethodAddParamInt(alpha);
            ScaleformMovieMethodAddParamBool(centered);
            EndScaleformMovieMethod();

            SetStreamedTextureDictAsNoLongerNeeded(textureDict);

            minimaps.Add(minimaps.Count + 1, new(textureDict, textureName));
            return minimaps.Count;
        }

        /// <summary>
        /// Remove an overlay from the ingame Map
        /// </summary>
        /// <param name="overlayId">Id of the overlay to remove</param>
        public static async void RemoveOverlayFromMinimap(int overlayId)
        {
            if (overlay == 0) await Load();
            CallMinimapScaleformFunction(overlay, "REM_OVERLAY");
            ScaleformMovieMethodAddParamInt(overlayId);
            EndScaleformMovieMethod();
            minimaps.Remove(minimaps.Keys.ToList()[overlayId]);
        }
    }
}

File: ScaleformUI_V8/src/scaleforms/minimap/minimap.ts
import {Singleton} from "../../helpers/singleton.decorator";
import {noop, waitUntilReturns} from "../../helpers/loaders";
import {ScaleformHandler} from "../scaleform";

@Singleton
export class MinimapOverlays {
    private overlay = 0
    private readonly minimaps: {txd: string, txn: string}[] = []
    public async load() {
        this.overlay = AddMinimapOverlay("files/MINIMAP_LOADER.gfx")
        await waitUntilReturns(noop, () => HasMinimapOverlayLoaded(this.overlay), [1, true], 0)
        SetMinimapOverlayDisplay(this.overlay, 0.0, 0.0, 100.0, 100.0, 100.0)
    }

    public async addSizedOverlay(
        textureDict: string, textureName: string,
        x: number, y: number,
        rotation = 0, width = -1,
        height = -1, alpha = 100, centered = false) {

        if (!HasStreamedTextureDictLoaded(textureDict)) {
            await waitUntilReturns(
                () => RequestStreamedTextureDict(textureDict, false),
                () => HasStreamedTextureDictLoaded(textureDict),
                [1, true],
                0)
        }
        CallMinimapScaleformFunction(this.overlay, "ADD_SIZED_OVERLAY")
        ScaleformMovieMethodAddParamTextureNameString(textureDict)
        ScaleformMovieMethodAddParamTextureNameString(textureName)
        ScaleformMovieMethodAddParamFloat(x)
        ScaleformMovieMethodAddParamFloat(y)
        ScaleformMovieMethodAddParamFloat(Math.round(rotation * 100) / 100)
        ScaleformMovieMethodAddParamFloat(Math.round(width * 100) / 100)
        ScaleformMovieMethodAddParamFloat(Math.round(height * 100) / 100)
        ScaleformMovieMethodAddParamInt(alpha)
        ScaleformMovieMethodAddParamBool(centered)
        EndScaleformMovieMethod()

        SetStreamedTextureDictAsNoLongerNeeded(textureDict)
        return this.minimaps.push({txd: textureDict, txn: textureName}) - 1
    }

    public async addScaledOverlay(
        textureDict: string, textureName: string,
        x: number, y: number,
        rotation = 0, xScale = 100.0,
        yScale = 100.0, alpha = 100, centered = false
    ): Promise<number> {

        if (!HasStreamedTextureDictLoaded(textureDict)) {
            await waitUntilReturns(
                () => RequestStreamedTextureDict(textureDict, false),
                () => HasStreamedTextureDictLoaded(textureDict),
                [1, true],
                0)
        }

        CallMinimapScaleformFunction(this.overlay, "ADD_SCALED_OVERLAY")
        ScaleformMovieMethodAddParamTextureNameString(textureDict)
        ScaleformMovieMethodAddParamTextureNameString(textureName)
        ScaleformMovieMethodAddParamFloat(x)
        ScaleformMovieMethodAddParamFloat(y)
        ScaleformMovieMethodAddParamFloat(Math.round(rotation * 100) / 100)
        ScaleformMovieMethodAddParamFloat(Math.round(xScale * 100) / 100)
        ScaleformMovieMethodAddParamFloat(Math.round(yScale * 100) / 100)
        ScaleformMovieMethodAddParamInt(alpha)
        ScaleformMovieMethodAddParamBool(centered)
        EndScaleformMovieMethod()

        SetStreamedTextureDictAsNoLongerNeeded(textureDict)
        return this.minimaps.push({ txd: textureDict, txn: textureName }) - 1
    }

    public async removeOverlayById(id: number) {
        if (id == 0) await this.load()
        CallMinimapScaleformFunction(this.overlay, "REM_OVERLAY");
        ScaleformMovieMethodAddParamInt(id)
        EndScaleformMovieMethod()
        delete this.minimaps[id]
    }
}